src/plots/mapbox/mapbox.js

```diff
@@ -111,125 +114,56 @@ proto.createMap = function(calcData, fullLayout, resolve, reject) {
 
     self.rejectOnError(reject);
 
-    map.once('load', function() {
-        self.updateData(calcData);
-        self.updateLayout(fullLayout);
-        self.resolveOnRender(resolve);
-    });
-
-    if(self.isStatic) return;
-
-    var wheeling = false;
-
-    // keep track of pan / zoom in user layout and emit relayout event
-    map.on('moveend', function(eventData) {
-        if(!self.map) return;
-
-        // 'moveend' gets triggered by map.setCenter, map.setZoom,
-        // map.setBearing and map.setPitch.
-        //
-        // Here, we make sure that state updates amd 'plotly_relayout'
-        // are triggered only when the 'moveend' originates from a
-        // mouse target (filtering out API calls) to not
-        // duplicate 'plotly_relayout' events.
-
-        if(eventData.originalEvent || wheeling) {
-            var optsNow = gd._fullLayout[self.id];
-            Registry.call('_storeDirectGUIEdit', gd.layout, gd._fullLayout._preGUI, self.getViewEdits(optsNow));
-
-            var viewNow = self.getView();
-            optsNow._input.center = optsNow.center = viewNow.center;
-            optsNow._input.zoom = optsNow.zoom = viewNow.zoom;
-            optsNow._input.bearing = optsNow.bearing = viewNow.bearing;
-            optsNow._input.pitch = optsNow.pitch = viewNow.pitch;
-
-            gd.emit('plotly_relayout', self.getViewEdits(viewNow));
-        }
-        wheeling = false;
-    });
-
-    map.on('wheel', function() {
-        wheeling = true;
-    });
-
-    map.on('mousemove', function(evt) {
-        var bb = self.div.getBoundingClientRect();
+    if(!self.isStatic) {
+        self.initFx(calcData, fullLayout);
+    }
 
-        // some hackery to get Fx.hover to work
-        evt.clientX = evt.point.x + bb.left;
-        evt.clientY = evt.point.y + bb.top;
+    var promises = [];
 
-        evt.target.getBoundingClientRect = function() { return bb; };
+    promises.push(new Promise(function(resolve) {
+        map.once('load', resolve);
+    }));
 
-        self.xaxis.p2c = function() { return evt.lngLat.lng; };
-        self.yaxis.p2c = function() { return evt.lngLat.lat; };
+    promises = promises.concat(self.fetchMapData(calcData, fullLayout));
 
-        Fx.hover(gd, evt, self.id);
-    });
+    Promise.all(promises).then(function() {
+        self.updateData(calcData);
+        self.updateLayout(fullLayout);
+        self.resolveOnRender(resolve);
+    }).catch(reject);
+};
 
-    function unhover() {
-        Fx.loneUnhover(fullLayout._toppaper);
+proto.fetchMapData = function(calcData) {
+    var promises = [];
+
+    function fetch(url) {
+        return new Promise(function(resolve, reject) {
+            d3.json(url, function(err, d) {
+                if(err) {
+                    delete PlotlyGeoAssets[url];
+                    var msg = err.status === 404 ?
+                        ('GeoJSON at URL ' + url + ' does not exist.') :
```

Comment:
Adding double quotations before & after url may help the message to be slightly more readable.

Reply:
done in https://github.com/plotly/plotly.js/pull/3988/commits/e509047a6c99db6f2e452d20c304587ce51ce1cb